Add httptrace based prober

discordianfish opened the pull request:
Hi,

this is pretty much the prober I use for the latency.at exporters. I've added it as a new module here although in my exporters I replaced the http module which might make sense, given that it basically only adds one new metric.

Beside that I still need refactor and move over some tests. But wanted to get feedback early. I expected especially the redirect stuff to something to discuss.

brian-brazil commented:
Yeah, it doesn't make sense to have this as a new module. Metrics should Just Work from the user standpoint.
My inclination would be not to use labels for these, as the existing dns metric doesn't have them and it'd make us less likely to forget to set one.
We should mention in the help strings how this interacts with redirects, e.g. the stats are for the first request.

discordianfish commented:
@brian-brazil I've change the PR to extend the existing http probe. From a user perspective it should be just a new metric that appears. I also simplified the tracing a bit.

Which labels do you mean? The duration labels should be fine, given it's a new metrics, right?
For the SSL metrics in the redirect we could either keep it as it is (which is stats for *last* request though, or do I miss something?) and document it in the help string, or we could add the redirect number label too but it would be a breaking change. I don't feel strongly either way.

brian-brazil commented:
> Which labels do you mean? The duration labels should be fine, given it's a new metrics, right?

I mean the labels on durationGaugeVec so that it matches the existing dns duration metric, no need to duplicate that information and cause confusion.

> which is stats for last request though, or do I miss something?

Yes, it's the last request we get there as that's what the library gives us.

brian-brazil commented on prober/http.go:
```diff
@@ -128,7 +200,7 @@ func ProbeHTTP(ctx context.Context, target string, module config.Module, registr
 	}
 
-	ip, err := chooseProtocol(module.HTTP.PreferredIPProtocol, targetHost, registry, logger)
+	ip, err := chooseProtocolMetrics(module.HTTP.PreferredIPProtocol, targetHost, durationGaugeVec.WithLabelValues("resolve"), probeIPProtocolGauge, logger)
```
This removes the existing duration gauge, which needs to stay around. If you need the duration, get the function to pass it back rather than splitting the logic in two and being inconsistent.

discordianfish replied:
Yes, the idea was to use the http durationGauge with 'resolve' label for that instead.
I could expose both?
I definitely would want to have the resolution in the time in the overall http duration metric. It's just so much more convienient to consume.

brian-brazil commented on prober/http.go:
```diff
@@ -65,9 +67,73 @@ func matchRegularExpressions(reader io.Reader, httpConfig config.HTTPProbe, logg
 	var (
+		durationGaugeVec = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Name: "duration_seconds",
+			Help: "Duration of http request by phase, summed over all redirects",
```
if it's http then http should be in the metric name

brian-brazil commented on prober/http.go:
```diff
@@ -243,6 +330,38 @@ func ProbeHTTP(ctx context.Context, target string, module config.Module, registr
+		i                   = 0
+	)
+	for _, trace := range tt.traces {
+		// We get the duration for the first request from chooseProtocol
```
It's missing that from here then, so isn't giving a full view

discordianfish replied:
Not sure what you mean. At this point we already recorded the time to resolve the first redirect in chooseProtocolMetrics, so that should be right, no?

brian-brazil commented on prober/http.go:
```diff
@@ -243,6 +330,38 @@ func ProbeHTTP(ctx context.Context, target string, module config.Module, registr
+		earlierstCertExpiry time.Time
+		i                   = 0
+	)
+	for _, trace := range tt.traces {
```
This should all be logged

brian-brazil commented on prober/http.go:
```diff
@@ -243,6 +330,38 @@ func ProbeHTTP(ctx context.Context, target string, module config.Module, registr
+			durationGaugeVec.WithLabelValues("resolve").Add(trace.dnsDone.Sub(trace.start).Seconds())
+		}
+		if resp.TLS != nil {
+			exp := getEarliestCertExpiry(resp.TLS)
```
This is conflicting with the logic below, as there's two different sets of TLS expiration logic. Keep just the existing one that looks only at the last request, for simplicity

discordianfish replied:
Oh right, missed that when rebasing.

brian-brazil commented on prober/http.go:
```diff
@@ -243,6 +330,38 @@ func ProbeHTTP(ctx context.Context, target string, module config.Module, registr
+				success = false
+			}
+			durationGaugeVec.WithLabelValues("connect").Add(trace.connectDone.Sub(trace.dnsDone).Seconds())
+			durationGaugeVec.WithLabelValues("tls").Add(trace.gotConn.Sub(trace.dnsDone).Seconds())
```
This is going to be inconsistent with the name of the equivalent metric for the tcp probe

discordianfish replied:
Yes.. As I said, I feel like the http duration metric is with these labels much more convinient to use.
We could also make the tcp probe and all other probes put these timings into one duration gauge for consistency. What do you think about that? That would be a breaking change though.

brian-brazil replied:
I'm not sure what the best way forward is here, http having multiple requests messes things up a bit. Http phases don't also apply to other probes.

discordianfish replied:
> I'm not sure what the best way forward is here, http having multiple requests messes things up a bit

I'm already summing up the times for the redirects, so there is only one timeseries per phase.


> Http phases don't also apply to other probes.

Every probe has some sort of phases, right? They are even similar: resolve/connect/tls/processing/transfer should apply 1:1 to tcp probe. A subset of that would work for udp too. For ICMP we could still have resolve/waiting or something like that. DNS is tricky though given that Exchange() is pretty high level. But that could still use the same gauge with a single label.

In the end I'm fine with having 'inconsistent' metrics or just expose times like tls and dns twice.

brian-brazil replied:
> resolve/connect/tls/processing/transfer should apply 1:1 to tcp probe

Not quite, HTTP is a single query and response whereas TCP can have multiple queries and responses in what HTTP would consider "transfer".

discordianfish replied:
Oh right now I see what you mean, didn't use the tcp probes much.
There is probably no way to get something consistent here. I'm leaning towards just exposing the tls and dns timings twice.

brian-brazil replied:
That's probably the best route, we'll just need to be clear in the HELP strings about how they're different.

discordianfish commented:
Okay, I think I addressed all feedback. This also logs each request on debug level.

brian-brazil commented on prober/http.go:
```diff
@@ -65,9 +67,76 @@ func matchRegularExpressions(reader io.Reader, httpConfig config.HTTPProbe, logg
 	return true
 }
 
+// trace holds timings for a single http roundtrip.
```
HTTP

brian-brazil commented on prober/http.go:
```diff
@@ -65,9 +67,76 @@ func matchRegularExpressions(reader io.Reader, httpConfig config.HTTPProbe, logg
+	end           time.Time
+}
+
+// transport is a custom transport keeping traces for each http roundtrip.
```
Here too

brian-brazil commented on prober/http.go:
```diff
@@ -65,9 +67,76 @@ func matchRegularExpressions(reader io.Reader, httpConfig config.HTTPProbe, logg
+
+// RoundTrip switches to a new trace, then runs embedded RoundTripper.
+func (t *transport) RoundTrip(req *http.Request) (*http.Response, error) {
+	if _, ok := t.traces[req]; ok {
```
There's already a limit on the number of redirects, so we don't also need this.

brian-brazil commented on prober/http.go:
```diff
@@ -65,9 +67,76 @@ func matchRegularExpressions(reader io.Reader, httpConfig config.HTTPProbe, logg
+	}
+}
+func (t *transport) ConnectDone(net, addr string, err error) {
+	// t.current.addr = addr
```
Old debug code?

brian-brazil commented on prober/http.go:
```diff
@@ -146,6 +217,10 @@ func ProbeHTTP(ctx context.Context, target string, module config.Module, registr
 		return false
 	}
 
+	// Inject transport that tracks trace for each redirect
```
Full stop

brian-brazil commented on prober/http.go:
```diff
@@ -65,9 +67,76 @@ func matchRegularExpressions(reader io.Reader, httpConfig config.HTTPProbe, logg
+	}
+	t.traces[req] = &trace{}
+	t.current = t.traces[req]
+	level.Debug(t.logger).Log("msg", "Processing request", "url", req.URL.String())
```
We already have this logged from the redirect logic

brian-brazil commented on prober/http.go:
```diff
@@ -65,9 +67,76 @@ func matchRegularExpressions(reader io.Reader, httpConfig config.HTTPProbe, logg
+}
+func (ts *transport) ConnectStart(_, _ string) {
+	t := ts.current
+	// No DNS resolution, e.g connecting to IP directly
```
e.g.

brian-brazil commented on prober/http.go:
```diff
@@ -247,6 +333,26 @@ func ProbeHTTP(ctx context.Context, target string, module config.Module, registr
+		if i != 0 {
+			durationGaugeVec.WithLabelValues("resolve").Add(trace.dnsDone.Sub(trace.start).Seconds())
+		}
+		if resp.TLS != nil {
```
Just because the last response wasn't TLS doesn't mean previous ones were, you always should add in tls

discordianfish replied:
Oh good catch. I think we need to track which trace involved TLS.

brian-brazil commented on prober/http.go:
```diff
@@ -247,6 +333,26 @@ func ProbeHTTP(ctx context.Context, target string, module config.Module, registr
+			durationGaugeVec.WithLabelValues("tls").Add(trace.gotConn.Sub(trace.dnsDone).Seconds())
+		} else {
+			durationGaugeVec.WithLabelValues("connect").Add(trace.gotConn.Sub(trace.dnsDone).Seconds())
+			if httpConfig.FailIfNotSSL {
```
This logic is already handled below

brian-brazil commented on prober/utils.go:
```diff
@@ -11,9 +11,8 @@ import (
 )
 
 // Returns the preferedIPProtocol, the dialProtocol, and sets the probeIPProtocolGauge.
```
This comment needs updating

discordianfish commented:
Note to myself: Looks like something is broken with the trace metrics for IPv6:
```
```
level=info msg="Resolving target address" preferred_ip_protocol=ip6
level=info msg="Resolved target address" ip=2a00:1450:4016:80a::200e
level=info msg="Making HTTP request" url=http://[2a00:1450:4016:80a::200e] host=google.com
level=debug msg="Processing request" url=http://[2a00:1450:4016:80a::200e]
level=error msg="Error for HTTP request" err="Get http://[2a00:1450:4016:80a::200e]: dial tcp [2a00:1450:4016:80a::200e]:80: connect: network is unreachable"
```

```
probe_http_duration_seconds{phase="connect"} -9.223372036854776e+09
probe_http_duration_seconds{phase="processing"} 0
probe_http_duration_seconds{phase="resolve"} 0.021426744
probe_http_duration_seconds{phase="transfer"} 9.223372036854776e+09
```

discordianfish commented:
I've addressed the comments but still need to look into the IPV6 issue.

discordianfish commented:
@brian-brazil How is the chooseProtocol logic suppose to work?

It looks like by default it sets
```
                preferredIPProtocol = "ip6"
                fallbackProtocol = "ip4"
```

in which case it just resolves the name and returns it. I expected it somehow checks whether IPv6 is working? I guess that's as intended and people need to explicitly set preferred protocol to v4 if their hosts have no IPv6 support?

brian-brazil commented:
> I guess that's as intended and people need to explicitly set preferred protocol to v4 if their hosts have no IPv6 support?

Yes, that's how it works.

discordianfish commented:
The problem with ipv6 was due to the traces being empty when the request failed. I'm not skipping traces where gotConn is zero.

brian-brazil commented:
What if there's resolution time and then the request fails?

discordianfish commented:
@brian-brazil Not sure if you saw the earlier version but I moved the skipping below recording of resolution, so you would still get resolution times even if the request itself fails.
Arguably the request could also fail later and it would make sense to expose timings for connect etc but I feel that's not worth the more complicated code. But if you feel different, I can look into doing that.

I've also added a test. This only checks if the metric+values exist though. I don't think that sleeping and fuzzy matching the timings would be the right way. What do you think?

brian-brazil commented on prober/http.go:
```diff
@@ -65,9 +67,75 @@ func matchRegularExpressions(reader io.Reader, httpConfig config.HTTPProbe, logg
+type transport struct {
+	Transport http.RoundTripper
+	logger    log.Logger
+	traces    map[*http.Request]*trace
```
This should be a list rather than a map

discordianfish replied:
Good point, I've changed it.

brian-brazil commented on prober/http.go:
```diff
@@ -65,9 +67,75 @@ func matchRegularExpressions(reader io.Reader, httpConfig config.HTTPProbe, logg
+}
+func (ts *transport) ConnectStart(_, _ string) {
+	t := ts.current
+	// No DNS resolution, e.g. connecting to IP directly
```
Full stop at the end of comments

brian-brazil commented on prober/http.go:
```diff
@@ -247,6 +332,27 @@ func ProbeHTTP(ctx context.Context, target string, module config.Module, registr
 	}
+	i := 0
+	for _, trace := range tt.traces {
+		// We get the duration for the first request from chooseProtocol
```
You should log the times too, so the user can get back to the individual request times

discordianfish replied:
I've added logging the raw timestamps which is probably most useful.

brian-brazil replied:
I think the diffs would be more useful to a human.

discordianfish commented:
@brian-brazil Okay, so from my perspective this is ready to get merged. Let me know if there is more to change, preferably in one batch so I can address all remaining changes at once.

discordianfish commented:
Okay hold off on merging this. Look like we have similar issues with the other timings: A connection can break at any point, so we always need to check if a timing is available or not. Will update the PR in a bit.

Rudd-O commented:
This looks quite useful.

discordianfish commented:
I think this should do it. I've deployed it here and will report back if things look good in prod too.

discordianfish commented:
I've been running this for the past 20 days and things look good. Going to rebase now. Can we then get this merged?

brian-brazil commented on prober/http.go:
```diff
@@ -65,9 +67,76 @@ func matchRegularExpressions(reader io.Reader, httpConfig config.HTTPProbe, logg
 	return true
 }
 
+// trace holds timings for a single HTTP roundtrip.
```
roundTripTrace

brian-brazil commented on prober/http.go:
```diff
@@ -65,9 +67,76 @@ func matchRegularExpressions(reader io.Reader, httpConfig config.HTTPProbe, logg
+}
+func (ts *transport) ConnectStart(_, _ string) {
+	t := ts.current
+	// No DNS resolution, e.g. connecting to IP directly.
```
s/e.g.//

e.g. means "for example" and this is the only case where there'd be no DNS

brian-brazil commented on prober/http.go:
```diff
@@ -247,6 +333,42 @@ func ProbeHTTP(ctx context.Context, target string, module config.Module, registr
+			continue
+		}
+		if trace.tls {
+			// dnsDone must be set if gotConn was set
```
Comments end with a full stop

brian-brazil commented on prober/http.go:
```diff
@@ -65,9 +67,76 @@ func matchRegularExpressions(reader io.Reader, httpConfig config.HTTPProbe, logg
 func ProbeHTTP(ctx context.Context, target string, module config.Module, registry *prometheus.Registry, logger log.Logger) (success bool) {
 	var redirects int
 	var (
+		durationGaugeVec = prometheus.NewGaugeVec(prometheus.GaugeOpts{
```
We should initialise this with all the potential phases.

discordianfish replied:
How?

brian-brazil replied:
Call WithLabelValues with all the potential values.

discordianfish replied:
Like this?

brian-brazil commented on prober/http.go:
```diff
@@ -247,6 +333,42 @@ func ProbeHTTP(ctx context.Context, target string, module config.Module, registr
 	if resp == nil {
 		resp = &http.Response{}
 	}
+	i := 0
```
You don't need a separate i now that tt.traces is a list, it's the first variable from the range.

discordianfish replied:
Good catch. Still a left over from earlier where this wasn't needed.

brian-brazil commented on prober/utils.go:
```diff
@@ -63,5 +62,7 @@ func chooseProtocol(preferredIPProtocol string, target string, registry *prometh
 	level.Info(logger).Log("msg", "Resolved target address", "ip", ip)
-	return ip, nil
+	lookupTime := time.Since(resolveStart).Seconds()
+	probeDNSLookupTimeSeconds.Add(lookupTime)
```
This is going to double count with the defer above. We should just have the defer, and can use a named return value for it.

discordianfish commented:
Ok, made these changes except the initializing. Not sure what you mean by that one..?

brian-brazil commented on prober/http.go:
```diff
@@ -173,6 +173,11 @@ func ProbeHTTP(ctx context.Context, target string, module config.Module, registr
 		})
 	)
 
+	// Initialize durationGaugeVec
```
You don't really need this comment.

discordianfish commented:
Okay, removed the comment. Ready to merge now?
